Add pluggable Logger for pipeline

bobheadxi opened the pull request:
closes #244 

This PR allows something like the following: (*edit* no longer possible as of https://github.com/src-d/hercules/pull/262/commits/c92aaacc8b5498a6d9f9b5699a2aed9ea904f85b - a wrapper needs to be written around zap)

```go
package hercules_test

import (
	"go.uber.org/zap"
	gogit "gopkg.in/src-d/go-git.v4"
	"gopkg.in/src-d/hercules.v10"
)

func main() {
	repo, _ := gogit.PlainOpen(".")
	pipe := hercules.NewPipeline(repo)
	pipe.Initialize(map[string]interface{}{
                // custom logger
		hercules.ConfigLogger: zap.NewExample().Sugar(),
	})
}
```

I've added a `Logger` and `Config` variable for `facts` in package `internal/core`, and also added the logger to all pipeline items and replaced the log calls that i could find.

There are still a few calls to `panic` that I think we should try to remove in favour of `error`, but I noticed some tests actually assert that panics occur so I've tried not to change too much for now (notably `BurndownAnalysis`)

bobheadxi commented:
@vmarkovtsev there appears to be an issue with `labours` even though nothing outward-facing has changed with this PR - do you have any ideas as to why this might be happening?

https://travis-ci.com/src-d/hercules/jobs/188673981#L2785

vmarkovtsev commented:
Yes, the problem is with Google's `ortools`, which is a dependency of `seriate`, which is a dependency of `labours`. The released v7 on March 26th which is not backward compatible, and I did not specify the upper version constraint in `seriate`. Fixing this now.

bobheadxi commented:
Thanks @vmarkovtsev looks like the build is passing now!

vmarkovtsev commented:
Reviewing this tomorrow :+1:

vmarkovtsev commented on cmd/hercules/plugin.template:
```diff
@@ -81,11 +83,17 @@ func ({{.varname}} *{{.name}}) Description() string {
 // Initialize resets the internal temporary data structures and prepares the object for Consume().
 func ({{.varname}} *{{.name}}) Initialize(repository *git.Repository) error {
+  if {{.varname}}.l == nil {
+    {{.varname}}.l = hercules.NewLogger()
```
Indentation seems inconsistent here and above at line 87

vmarkovtsev commented on cmd/hercules/plugin.template:
```diff
@@ -81,11 +83,17 @@ func ({{.varname}} *{{.name}}) Description() string {
 
 // Initialize resets the internal temporary data structures and prepares the object for Consume().
 func ({{.varname}} *{{.name}}) Initialize(repository *git.Repository) error {
+  if {{.varname}}.l == nil {
```
We should always overwrite anything existing in `Initialize()`, that's the method's philosophy

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// NewLogger returns a configured default logger.
+func NewLogger() *DefaultLogger {
+	return &DefaultLogger{
+		I: log.New(os.Stdout, "[INFO] ", log.LstdFlags),
```
We should log everything to stderr by default, since stdout is used to return analysis results.

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+	}
+}
+
+// Info writes to info logger
```
```suggestion
// Info writes to "info" logger.
```

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// Info writes to info logger
+func (d *DefaultLogger) Info(v ...interface{}) { d.I.Print(v...) }
+
+// Infof writes to info logger
```
```suggestion
// Infof writes to "info" logger, printf-like format.
```

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// Infof writes to info logger
+func (d *DefaultLogger) Infof(f string, v ...interface{}) { d.I.Printf(f, v...) }
+
+// Warn writes to the warning logger
```
```suggestion
// Warn writes to the "warning" logger.
```

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// Warn writes to the warning logger
+func (d *DefaultLogger) Warn(v ...interface{}) { d.W.Print(v...) }
+
+// Warnf writes to the warning logger
```
```suggestion
// Warnf writes to the "warning" logger, printf-like format.
```

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// Warnf writes to the warning logger
+func (d *DefaultLogger) Warnf(f string, v ...interface{}) { d.W.Printf(f, v...) }
+
+// Error writes to the error logger
```
```suggestion
// Error writes to the "error" logger.
```

vmarkovtsev commented on internal/core/logger.go:
```diff
@@ -0,0 +1,54 @@
+// Error writes to the error logger
+func (d *DefaultLogger) Error(v ...interface{}) { d.E.Print(v...) }
+
+// Errorf writes to the error logger
```
```suggestion
// Errorf writes to the "error" logger, printf-like format.
```

vmarkovtsev commented on internal/core/pipeline.go:
```diff
@@ -565,7 +569,8 @@ func (pipeline *Pipeline) resolve(dumpPath string) {
 			if graph.AddEdge(key, name) == 0 {
-				log.Panicf("Unsatisfied dependency: %s -> %s", key, item.Name())
+				pipeline.l.Errorf("Unsatisfied dependency: %s -> %s", key, item.Name())
+				return
```
We should panic here, because this is really a bug and an assertion failure.

vmarkovtsev commented on internal/core/pipeline.go:
```diff
@@ -618,7 +623,8 @@ func (pipeline *Pipeline) resolve(dumpPath string) {
 	if !ok {
-		panic("Failed to resolve pipeline dependencies: unable to topologically sort the items.")
+		pipeline.l.Errorf("Failed to resolve pipeline dependencies: unable to topologically sort the items.")
+		return
```
This is another real panic

vmarkovtsev commented on internal/core/pipeline.go:
```diff
@@ -788,14 +802,16 @@ func (pipeline *Pipeline) Run(commits []*object.Commit) (map[LeafPipelineItem]in
 					if !ok {
-						log.Panicf("%s: Consume() did not return %s", item.Name(), key)
+						err := fmt.Errorf("%s: Consume() did not return %s", item.Name(), key)
+						pipeline.l.Error(err)
```
This is another panic

vmarkovtsev commented on internal/core/pipeline_test.go:
```diff
@@ -460,11 +460,12 @@ func TestPipelineDeps(t *testing.T) {
 		"af9ddc0db70f09f3f27b4b98e415592a7485171c"))
 	result, err := pipeline.Run(commits)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
```
Oh jeez so many places check errors like the old version...

vmarkovtsev commented on internal/plumbing/renames.go:
```diff
@@ -384,9 +389,8 @@ func (ra *RenameAnalysis) blobsAreClose(blob1 *CachedBlob, blob2 *CachedBlob) (b
-			log.Println()
-			log.Println(blob1.Hash.String())
-			log.Println(blob2.Hash.String())
+			ra.l.Warnf("unclean return detected for blobs '%s' and '%s'\n",
```
```suggestion
			ra.l.Warnf("\nunclean return detected for blobs '%s' and '%s'\n",
```
because of the progress bar

vmarkovtsev commented on internal/plumbing/ticks.go:
```diff
@@ -120,8 +125,7 @@ func (ticks *TicksSinceStart) Consume(deps map[string]interface{}) (map[string]i
 		if ticks.tick0.Unix() < 631152000 { // 01.01.1990, that was 30 years ago
-			log.Println()
-			log.Printf("Warning: suspicious committer timestamp in %s > %s: %d",
+			ticks.l.Warnf("suspicious committer timestamp in %s > %s: %d",
```
```suggestion
			ticks.l.Warnf("\nsuspicious committer timestamp in %s > %s: %d",
```

vmarkovtsev commented on internal/plumbing/tree_diff.go:
```diff
@@ -165,7 +170,9 @@ func (treediff *TreeDiff) Consume(deps map[string]interface{}) (map[string]inter
 	if !pass && treediff.previousCommit != plumbing.ZeroHash {
-		log.Panicf("%s > %s", treediff.previousCommit.String(), commit.Hash.String())
+		err := fmt.Errorf("%s > %s", treediff.previousCommit.String(), commit.Hash.String())
+		treediff.l.Error(err)
```
This is another real panic - bug in the code

vmarkovtsev commented on leaves/couples.go:
```diff
@@ -212,14 +217,18 @@ func (couples *CouplesAnalysis) Finalize() interface{} {
-			log.Panicf("cannot find file %s in commit %s: %v",
+			err := fmt.Errorf("cannot find file %s in commit %s: %v",
 				name, couples.lastCommit.Hash.String(), err)
+			couples.l.Error(err)
```
This is another panic - "impossible"

vmarkovtsev commented on leaves/research/typos.go:
```diff
@@ -198,7 +203,7 @@ func (tdb *TyposDatasetBuilder) Consume(deps map[string]interface{}) (map[string
 			pos := uast.PositionsOf(n.(nodes.Object))
 			if pos.Start() == nil {
-				log.Printf("repo %s commit %s file %s adds identifier %s with no position",
+				tdb.l.Infof("repo %s commit %s file %s adds identifier %s with no position",
```
```suggestion
				tdb.l.Warnf("repo %s commit %s file %s adds identifier %s with no position",
```

vmarkovtsev commented on leaves/research/typos.go:
```diff
@@ -211,7 +216,7 @@ func (tdb *TyposDatasetBuilder) Consume(deps map[string]interface{}) (map[string
 			pos := uast.PositionsOf(n.(nodes.Object))
 			if pos.Start() == nil {
-				log.Printf("repo %s commit %s file %s removes identifier %s with no position",
+				tdb.l.Infof("repo %s commit %s file %s removes identifier %s with no position",
```
```suggestion
				tdb.l.Warnf("repo %s commit %s file %s removes identifier %s with no position",
```

bobheadxi commented:
@vmarkovtsev real bugs should still propagate errors upwards for the caller to deal with - for example, using Hercules as a library in a web service, a user would *definitely* not want a fatal crash every time, but instead log and handle it safely

bobheadxi commented:
If done right, returning errors will do the same thing as panic, since it’ll return up to the caller (for example, the Hercules command) and a panic can be done there, or simply exit with 1. Logger can be configured to capture stacktraces no errors as well

vmarkovtsev commented:
This is the thing, the places which I marked (I did not mark everything) are assertions which should *never* happen. The same as C++ assert or accessing a slice element out of bounds. If Go had macros, these would be disabled at all in the release build.

bobheadxi commented:
Hm... even then I believe it still better to propagate it properly, since sometimes unrelated things can cause certain impossibilities to happen, even if it doesn’t seem like it (at least it seems to be what happened here: https://github.com/src-d/hercules/pull/234#issuecomment-471200837)

Also feels good for the sake of consistency - out of bounds is one thing, but an unexpected invariant change feels like it should be returned properly at least based on what I understand of Go best practices, for the cases I mentioned earlier (in the event something goes wrong, even if it’s probably impossible, you want to give the caller the chance to clean up properly after itself, which could be critical - a panic interrupting a database write is less than pleasant)

If you feel strongly about keeping the panics I suppose I can change them back, but I would personally strongly recommend moving towards reducing the panics in the library component of Hercules (would be very helpful for me 😅)

vmarkovtsev commented:
Regarding catching panics in a lib, there is always `recover()` at your service.

vmarkovtsev commented:
I know your pain, I used to suffer the same way when I used [enry](https://github.com/src-d/enry) to detect languages, and it used to be very buggy and panic all the time when used concurrently. After all, I could report proper panic stack traces to the maintainers, and they were able to fix problems very fast. I do understand that we can change those logging calls to include stack traces. If you know how to do it (print a fancy stack trace when a 🐛 appears), to me it is the same as panicking. So I am fine with that actually.

bobheadxi commented:
Thanks! I actually did not consider recover at all, so thank you for pointing that out. I’ll look into both solutions (log stacktrace + return error vs panic) and see might be best for these cases, and make the necessary changes. I’ll apply fixes for your other comments as well.

Thanks for the review!

bobheadxi commented:
@vmarkovtsev I've added a stacktrace capture to all calls to `Error` and `Errorf`. The output looks like this:

```
[ERROR] 2019/04/01 17:08:50 hello-world
[ERROR] 2019/04/01 17:08:50 stacktrace:
runtime/debug.Stack(0xc000069e18, 0x10ce4e4, 0x480)
	/usr/local/Cellar/go/1.12/libexec/src/runtime/debug/stack.go:24 +0x9d
gopkg.in/src-d/hercules.v10/internal/core.captureStacktrace(0x0, 0x114ddce, 0xc000020320, 0x2)
	/Users/robertlin/go/src/gopkg.in/src-d/hercules.v10/internal/core/logger.go:76 +0x26
gopkg.in/src-d/hercules.v10/internal/core.(*DefaultLogger).logStacktraceToErr(0xc000069f60)
	/Users/robertlin/go/src/gopkg.in/src-d/hercules.v10/internal/core/logger.go:72 +0x2e
gopkg.in/src-d/hercules.v10/internal/core.(*DefaultLogger).Errorf(0xc000069f60, 0x16e5678, 0x5, 0xc000069f78, 0x2, 0x2)
	/Users/robertlin/go/src/gopkg.in/src-d/hercules.v10/internal/core/logger.go:62 +0x6f
gopkg.in/src-d/hercules.v10/internal/core.TestLogger(0xc000152300)
	/Users/robertlin/go/src/gopkg.in/src-d/hercules.v10/internal/core/logger_test.go:50 +0xbf4
testing.tRunner(0xc000152300, 0x170a120)
	/usr/local/Cellar/go/1.12/libexec/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.12/libexec/src/testing/testing.go:916 +0x357
```

does that look okay to you?

vmarkovtsev commented:
Yep, but let's make a separate level for such logs, e.g. "critical" or "fatal".

bobheadxi commented:
Hm, is there a particular reason why? My thoughts:

* adding `Critical()` breaks drop-in compatibility with many popular loggers (notably zap and logrus)
* adding `Fatal()` means we either follow standard behaviour of calling `os.Exit`, which means we're back to `panic`-like behaviour, or not calling `os.Exit` and introducing a behavioural inconsistency with popular loggers (zap, logrus)

Another thought: zap, by default, captures stacktraces on Error-level logs and above, so capturing stacktraces on `Error()` isn't out of the ordinary:

https://github.com/uber-go/zap/blob/master/config.go#L67

```go
	// DisableStacktrace completely disables automatic stacktrace capturing. By
	// default, stacktraces are captured for WarnLevel and above logs in
	// development and ErrorLevel and above in production.
	DisableStacktrace bool `json:"disableStacktrace" yaml:"disableStacktrace"`
```

Besides, a stacktrace should be interesting for all instances of `l.Error` right? Would be interested in hearing your thoughts on this

vmarkovtsev commented:
Most of the errors do not require printing the trace and the log will be dumped otherwise.

bobheadxi commented:
I've added `Critical` etc in https://github.com/src-d/hercules/pull/262/commits/2345fe706a31e35c7cd30e76716132899d5f322c

bobheadxi commented:
I've added some additional tests in daba29a

some drops are caused by the panic removals, where I've added a log and a return ([example](https://codecov.io/gh/src-d/hercules/compare/e03f209bf0e51450e9e82d96f0b144763eea4826...c92aaacc8b5498a6d9f9b5699a2aed9ea904f85b/diff#D5-x168), [example](https://codecov.io/gh/src-d/hercules/compare/e03f209bf0e51450e9e82d96f0b144763eea4826...c92aaacc8b5498a6d9f9b5699a2aed9ea904f85b/diff#D8-x568))
